uxarray/grid/validation.py

```diff
@@ -0,0 +1,44 @@
+import numpy as np
+from uxarray.io._ugrid import _is_ugrid
+from uxarray.constants import ERROR_TOLERANCE
+
+
+# validation helper functions
+def check_connectivity(self):
+    """Check if all nodes are referenced by at least one element.
+
+    If not, the mesh may have hanging nodes and may not a valid UGRID
+    mesh
+    """
+
+    # Check if all nodes are referenced by at least one element
+    # get unique nodes in connectivity
+    nodes_in_conn = np.unique(self.Mesh2_face_nodes.values.flatten())
+    #  remove negative indices/fill values from the list
+    nodes_in_conn = nodes_in_conn[nodes_in_conn >= 0]
+
+    # check if the size of unique nodes in connectivity is equal to the number of nodes
+    if (nodes_in_conn.size == self.nMesh2_node):
+        print("-All nodes are referenced by at least one element.")
+        return True
+    else:
+        print("-WARNING: Some nodes may not referenced by any element.",
+              nodes_in_conn.size, self.nMesh2_node)
+        return False
+
+
+def check_duplicate_nodes(self):
```

Comment:
This should be an internal method
```suggestion
def _check_duplicate_nodes(self):
```

Reply:
`self` is also typically reserved for use in classes, use `grid` instead.